### ISSUE TITLE
ErrorHandler does not throw on deprecations when configured so

### ISSUE THREAD

**Issue by ghost** [symfony/symfony#57551]
### Symfony version(s) affected

6.3.2

### Description

I set DI parameter `debug.error_handler.throw_at: 32767` (E_ALL) but deprecations do not throw. Why?

### How to reproduce

.

### Possible Solution

When I configure to throw on deprecations the please throw on deprecations.

**Comment by xabbuh** [comment]
Symfony 6.3 is not maintained anymore. Do you experience the same on 6.4?

**Comment by ghost** [comment]
I would have to create a reproducer because I'm not able to update my application on the fly.
However I'm pretty sure this is still valid issue: https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/ErrorHandler/ErrorHandler.php#L296C40-L296C118

**Comment by nicolas-grekas** [comment]
As you spotted out, this is by design. Deprecations shouldn't disrupt the code flow.

**Comment by nicolas-grekas** [comment]
Maybe https://symfony.com/doc/current/components/phpunit_bridge.html#display-the-full-stack-trace can help for your use case.

**Comment by ghost** [comment]
@nicolas-grekas please don't close the issue until both sides of the conversation reach an agreement.


> Deprecations shouldn't disrupt the code flow.

I would say `Deprecations shouldn't disrupt the code flow by default.`. Why framework arbitrary decides that deprecations should never throw? I have a use case when I want to throw.

Please reopen.

**Comment by nicolas-grekas** [comment]
It's not arbitrary - this matches the definition and the purpose of deprecations.
I don't know anything about the use case you have in mind so I can't comment on it.

**Comment by ghost** [comment]
> this matches the definition and the purpose of deprecations.

Ok, then to this by default but don't block overriding this.

The use case is that we are upgrading the app and we want to throw on deprecations during regression testing on QA.
I know there are workarounds but this the fastest way, it's one line change to add a DI parameter.

TL;DR just don't block overriding this. It just doesn't make sense to arbitrary do that.

**Comment by nicolas-grekas** [comment]
OK, then if testing, we built the phpunitbridge for that, I linked the corresponding doc page above.
It's not about blocking things but about providing structure.

**Comment by ghost** [comment]
By `regression testing on QA` I mean QA environments for manual QA team. This environments run with APP_ENV=prod and APP_DEBUG=1 set. Dev dependencies are not installed so phpunit does not exist (and wouldn't be used even if installed).

**Comment by nicolas-grekas** [comment]
Ah OK. Still, this is not a use case we'll want to support. What you can do is create a deprecation logger that throws instead of logging.

**Comment by ghost** [comment]
> What you can do is create a deprecation logger that throws instead of logging.

this proposal makes sense and should be fast enough to do. But could you elaborate more? You mean custom handler for monolog with channels configured to `deprecation`?

**Comment by nicolas-grekas** [comment]
Yes: the `deprecation` channel is already special: if you need it, deprecations will go there. Then, you'll have to configure this channel to make it throw.
